### PR TITLE
Allow copy & paste panels

### PR DESCRIPTION
**Pull request by swehner**
We use this mainly to diagnose problems we're having with our production site - while we have a lot of different dashboards,  sometimes it's very useful to quickly combine graphs from different dashboard to see all of them one after the other.
This makes it very easy to correlate different events.

I'm using the localStorage as a clipboard. To paste panels you can either add them to an existing row or create a new row with just this panel (what we're using mainly).

I don't have much experience with angular or javascript in general, so please let me know if anything's really weird or incorrect.

image

image

image

### CODE REVIEW

File: public/app/components/panelclipboard.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      var panel = null;
+      try {
+        panel = JSON.parse(json);
+      } catch(err) {
+        this.clear();
+      }
+      return panel;
+    },
+    set: function(panel) {
+      var copy = _.clone(panel);
```
Comment (craftytrickster) [inline public/app/components/panelclipboard.js]: Angular has a built in copy method already - [seen here](https://docs.angularjs.org/api/ng/function/angular.copy), that is more appropriate and would probably handle the need to manually delete the &&hashkey

Comment (swehner) [inline public/app/components/panelclipboard.js]: Thanks for the heads-up! I've changed it to what you suggested, and yes - it does seem to delete the &&hashKey automatically.

**Comment by torkelo** [comment]
I might be willing to merge this, but I think the copy action should be in the panel menu dropdown (the hamburger icon). Do not want to make the panel menu wider.

**Comment by swehner** [comment]
Okay, I don't like the burger menu much, because people will need one more click to get there...
Have you considered changing the actions in the list to icons? Something like this:?

image

I think the icons are pretty recognizable, and if they have a tooltip it should make it more clearer.
And the menu is more compact than before....
If you rather prefer it in the burger menu let me know and I'll move it there.

**Comment by torkelo** [comment]
Yea, I considered icons (think there are icons on some of the menu models still). Might be worth reconsidering.

**Comment by swehner** [comment]
Yeah, I've changed them a bit... don't know if it's the best way to do it, though...

Anyways, I'd love to get this integrated, so let me know if you'd rather have it in the burger menu, or in the main menu with the icons and I'll change it...

I'll add the patch in the comment so you can take a look.

```
diff --git a/public/app/components/panelmeta.js b/public/app/components/panelmeta.js
index 5c993fe..1ad536e 100644
--- a/public/app/components/panelmeta.js
+++ b/public/app/components/panelmeta.js
@@ -13,13 +13,13 @@ function () {
     this.extendedMenu = [];
 
     if (options.fullscreen) {
-      this.addMenuItem('view', 'icon-eye-open', 'toggleFullscreen(false); dismiss();');
+      this.addMenuItem('view', 'fa-eye', 'toggleFullscreen(false); dismiss();');
     }
 
-    this.addMenuItem('edit', 'icon-cog', 'editPanel(); dismiss();');
-    this.addMenuItem('duplicate', 'icon-copy', 'duplicatePanel()');
-    this.addMenuItem('copy', 'icon-copy', 'copyPanel(); dismiss();');
-    this.addMenuItem('share', 'icon-share', 'sharePanel(); dismiss();');
+    this.addMenuItem('edit', 'fa-edit', 'editPanel(); dismiss();');
+    this.addMenuItem('duplicate', 'fa-copy', 'duplicatePanel()');
+    this.addMenuItem('copy', 'fa-clipboard', 'copyPanel(); dismiss();');
+    this.addMenuItem('share', 'fa-share-square-o', 'sharePanel(); dismiss();');
 
     this.addEditorTab('General', 'app/partials/panelgeneral.html');
 
diff --git a/public/app/features/panel/panelMenu.js b/public/app/features/panel/panelMenu.js
index 27152ef..df76e0b 100644
--- a/public/app/features/panel/panelMenu.js
+++ b/public/app/features/panel/panelMenu.js
@@ -34,7 +34,12 @@ function (angular, $, _) {
           if (item.click) { template += ' ng-click="' + item.click + '"'; }
           if (item.editorLink) { template += ' dash-editor-link="' + item.editorLink + '"'; }
           template += '>';
-          template += item.text + '</a>';
+          if (item.icon) {
+            template += '<i class="fa ' + item.icon + '"/>';
+          } else {
+            template += item.text;
+          }
+          template += '</a>';
         });
 
         template += '</div>';
```

**Comment by torkelo** [comment]
Need to talk to the UIX and design colleagues that help me with Grafana and see what they think and if they have a better solution. 

ping @mattttt @bulletfactory

**Comment by mattttt** [comment]
Been thinking about this a bit, but need to sketch it out in greater detail. 

A couple early thoughts:
- Duplicate and copy panel are two versions of the same thing, just with destinations. I wonder if there is a way to consolidate that functionality somehow ('Duplicate: here' vs 'Duplicate in: Dashboard')

- Is it possible for the [ + Paste in New Row ] button to only appear if there is a panel in localStorage?

- Are you often deleting the new combo dashboards/rows after troubleshooting the issue? Or do they live on indefinitely?

**Comment by swehner** [comment]
The duplicate functionality is pretty much the same, you're right, I guess it depends a bit on how/if people use it often it might not justify having a separate button.

I'm not sure if the "Paste in row" button can appear if a panel is in localstorage - I'm not sure if this is easy to do with angular. The thing is that if you used "copy" once, the dashboard in the local storage will be there pretty much forever... (unless we have it time out somehow...)

We usually delete these dashboards afterwards - these are mostly to provide an easy way to comparing graphs and share them with other people...

**Comment by swehner** [comment]
Any news on this? Can I help in any other way to get this integrated?

**Comment by grzegorz-dubicki** [comment]
+1

**Comment by catacgc** [comment]
:+1:

**Comment by bulletfactory** [comment]
checking in on this.

**Comment by utkarshcmu** [comment]
+1 Are we merging this cool feature. I am sure its gonna be used a lot.

**Comment by keeganlow** [comment]
:+1: this would be massively helpful. @torkelo are the UX concerns still blocking merging this in? What are the open questions and/or action items that need to be addressed to move this forward? I'm happy to help in any way I can. Thanks!

**Comment by pakozm** [comment]
+1

**Comment by feraudet** [comment]
+1

**Comment by rgelobterFH** [comment]
+1

**Comment by Slach** [comment]
+1 
please merge this path ;) if it's possible

**Comment by akrus** [comment]
+1

**Comment by annapawlicka** [comment]
👍 👍 👍  this feature would be great!!

**Comment by swehner** [comment]
Hi again,

Seeing that there are a few more people interested in this and that the changes I had done on the 2.0 version didn't apply to master anymore, I've reapplied all changes.
Have you taken any decision on the UX issues @mattttt @torkelo ? Currently it's still on the main menu with the "Copy" but I'm happy to change it.
Let me know what you think.

**Comment by rajdeepbhattacharya** [comment]
Is this merged ? I see a lot of usage for me , can save a lot of time.
+1

**Comment by lucadistefano** [comment]
+1

**Comment by rajdeepbhattacharya** [comment]
@swehner @torkelo    : One more suggestion , if we can also implement duplicating a row in different dashboard as well. This will also help us calling different rows to a single dashbaord.

**Comment by andyxning** [comment]
any plans or point-in-time to merge this? it will be a productive functionality. 👍

**Comment by pokemane** [comment]
+1 for time savings and eliminating errors introduced through manual copying.

**Comment by Will-Beninger** [comment]
This is taking up a huge amount of my time presently, can this get merged?

It's been over a year since the first pull request....

**Comment by keen99** [comment]
+1 I'm doing this multiple times a day as well, and it doesnt help that I can't command-A to select all in the json panel....

**Comment by DebianRoxx** [comment]
This is a very nice feature ! :+1: 
We're generate automaticaly server's dashboards and sometimes, we want homemade dashboard just by clic & paste, actually it's not possible easily.

**Comment by BrianGallew** [comment]
For the love of all that's holy, instead of irritating the community trying to reach a poorly-defined "perfect",  could you merge this already?  The original diff has already been amended once to try to remain relevant.

**Comment by melor** [comment]
+1 Please get this merged. Thanks!

**Comment by utkarshcmu** [comment]
I am not sure when will this PR be merged or if there can be a better PR than this for the same use case. But I am writing an efficient Command Line tool named 'wizzy' since last 3-4 days which is ready to move your rows and panels across the dashboards and lets your store the dashboards in a Git repo as well. 

https://utkarshcmu.github.io/wizzy/
https://github.com/utkarshcmu/wizzy

This tool is still in its initial stages and needs a lot of work before we even call the code modular production ready. But, IT WORKS! Feel free to open issues or send a PR on wizzy repo!

**Comment by utkarshcmu** [comment]
Added some copy and move commands to wizzy - https://utkarshcmu.github.io/wizzy/
Now, we can move or copy panels and rows between dashboards easily using wizzy. Feedback appreciated. :)

**Comment by freimanas** [comment]
cant wait for this feature to be merged in!

**Comment by slnowak** [comment]
Is it already implemented in the latest grafana version? If not, will this PR be ever merged? What's the reason it hangs here for 1,5 years?

**Comment by torkelo** [comment]
Copy pasting panels is a great feature and something Grafana needs, just need to find the right UX for it. 

This PR did not find a good enough UX solution and has no also drifted too far from master to be in a mergable state. 

If you have ideas for how to do this from UX perspective please post ideas on the feature request:
https://github.com/grafana/grafana/issues/1004

**Comment by luisdavim** [comment]
@torkelo I agree the UX could be better but it's worst UX not having this feature, I think it would be much better to accept PRs that implement missing features and then working on improving them than to just discard things like this.
I feel this project is suffering a lot from this pattern of just rejecting contributions, this discourages people from trying to help.

**Comment by torkelo** [comment]
Sorry you feel that way. I do understand and agree that there is a lot to improve in Grafana in terms of features, ux, easy of use, simple common work flows etc. 

I realize that having a high standard for PR can  be bad and discourage contributions. But what can also kill projects are too many features, with a cluttered ux that caters to everyone's edge case scenarios. Features and ux actions needs a strong reason to exist and be implemented in a way that doesn't  distraction to much from other common actions and the new user experience / easy of use. 

It's easy to accept PRs and add new features but very hard to remove unwanted / unused features, this is another reason why some PRs do not get merged, or take time before they are merged.

Grafanas dev team has the last 3 years been very small (1-2) so working on PRs and moving the project vision / strategy forward has been challenging. We are expanding the team and will be able to spend more time on PRs in the future

**Comment by slnowak** [comment]
I understand your concerns and kind of agree with your approach, but this doesn't look like something unwanted/unused and could improve day to day Grafana usage even if it has crappy design.

**Pull request closed without merging**